pkg/cmd/run/watch/watch.go

```diff
@@ -0,0 +1,228 @@
+
+	if opts.Prompt {
+		cs := opts.IO.ColorScheme()
+		runs, err := shared.GetRunsWithFilter(client, repo, 10, func(run shared.Run) bool {
+			return run.Status != shared.Completed
+		})
+		if err != nil {
+			return fmt.Errorf("failed to get runs: %w", err)
+		}
+		if len(runs) == 0 {
+			return fmt.Errorf("found no in progress runs to watch")
+		}
+		runID, err = shared.PromptForRun(cs, runs)
+		if err != nil {
+			return err
+		}
+		// TODO silly stopgap until dust settles and PromptForRun can just return a run
+		for _, r := range runs {
+			if fmt.Sprintf("%d", r.ID) == runID {
+				run = &r
+				break
+			}
+		}
+	} else {
+		run, err = shared.GetRun(client, repo, runID)
+		if err != nil {
+			return fmt.Errorf("failed to get run: %w", err)
+		}
+	}
+
+	if run.Status == shared.Completed {
+		return nil
+	}
+
+	prNumber := ""
+	number, err := shared.PullRequestForRun(client, repo, *run)
+	if err == nil {
+		prNumber = fmt.Sprintf(" #%d", number)
+	}
+
+	if runtime.GOOS == "windows" {
+		opts.IO.EnableVirtualTerminalProcessing()
+	}
+	// clear entire screen
+	fmt.Fprintf(opts.IO.Out, "\x1b[2J")
+
+	annotationCache := map[int][]shared.Annotation{}
+
+	duration, err := time.ParseDuration(fmt.Sprintf("%ds", opts.Interval))
+	if err != nil {
+		return fmt.Errorf("could not parse interval: %w", err)
+	}
+
+	for run.Status != shared.Completed {
+		run, err = renderRun(*opts, client, repo, run, prNumber, annotationCache)
+		if err != nil {
+			return err
+		}
+		time.Sleep(duration)
+	}
+
+	if opts.ExitStatus && run.Conclusion != shared.Success {
+		return cmdutil.SilentError
+	}
+
+	return nil
+}
+
+func renderRun(opts WatchOptions, client *api.Client, repo ghrepo.Interface, run *shared.Run, prNumber string, annotationCache map[int][]shared.Annotation) (*shared.Run, error) {
+	out := opts.IO.Out
+	cs := opts.IO.ColorScheme()
+
+	var err error
+
+	run, err = shared.GetRun(client, repo, fmt.Sprintf("%d", run.ID))
+	if err != nil {
+		return run, fmt.Errorf("failed to get run: %w", err)
+	}
+
+	ago := opts.Now().Sub(run.CreatedAt)
+
+	jobs, err := shared.GetJobs(client, repo, *run)
+	if err != nil {
+		return run, fmt.Errorf("failed to get jobs: %w", err)
+	}
+
+	var annotations []shared.Annotation
+
+	var annotationErr error
+	var as []shared.Annotation
+	for _, job := range jobs {
+		if as, ok := annotationCache[job.ID]; ok {
+			annotations = as
+			continue
+		}
+
+		as, annotationErr = shared.GetAnnotations(client, repo, job)
+		if annotationErr != nil {
+			break
+		}
+		annotations = append(annotations, as...)
+
+		if job.Status != shared.InProgress {
+			annotationCache[job.ID] = annotations
+		}
+	}
+
+	if annotationErr != nil {
+		return run, fmt.Errorf("failed to get annotations: %w", annotationErr)
+	}
+
+	if runtime.GOOS == "windows" {
+		// Just clear whole screen; I wasn't able to get the nicer cursor movement thing working
+		fmt.Fprintf(opts.IO.Out, "\x1b[2J")
+	} else {
+		// Move cursor to 0,0
+		fmt.Fprint(opts.IO.Out, "\x1b[0;0H")
```

Comment:
In the future, when we have more time, I wonder whether we could keep track of the number of lines we've printed and move back and re-render only that amount. That way we keep the rest of the user's viewport intact

Reply:
that would be nice, but yeah, would take more time; it would also lead to visually jerking the viewport up as we exceed whatever space is initially available. Definitely something I'd like to experiment with for a hackday and do some comparative demos on.